sklearn/preprocessing/imputation.py

```diff
@@ -366,13 +397,53 @@ def transform(self, X):
 
             mask = _get_mask(X, self.missing_values)
             n_missing = np.sum(mask, axis=self.axis)
-            values = np.repeat(valid_statistics, n_missing)
 
-            if self.axis == 0:
-                coordinates = np.where(mask.transpose())[::-1]
+            if self.strategy == 'knn':
+                if self.axis == 1:
+                    X = X.transpose()
+                    mask = mask.transpose()
+                    statistics = statistics.transpose()
+                missing_index = np.where(mask.any(1))[0]
+                if False:
+                    for i, row in zip(missing_index, X[missing_index]):
+                        col_na_mask = np.isnan(row)
+                        col_full_mask = np.logical_not(col_na_mask)
+                        col_index = np.where(col_full_mask)[0]
+                        impute_index = np.where(col_na_mask)[0]
+                        neigh = NearestNeighbors(self.n_neighbors)
+                        neigh = neigh.fit(statistics[:, col_index])
+                        _dist, ind = neigh.kneighbors(row[col_full_mask],
+                                                       self.n_neighbors)
+                        nn_index = ind[0]
+                        X[i][impute_index] = statistics[nn_index][:, impute_index].mean(0)
+                else:
+
+                    #@jnothman 's method
+                    for sl in list(gen_batches(len(missing_index),100)):
+                        index_start, index_stop = missing_index[sl][0],missing_index[sl][-1]+1
+                        X_sl = X[index_start: index_stop].copy()
+                        mask_sl = _get_mask(X_sl, self.missing_values)
+                        missing_index_sl = np.where(mask_sl.any(1))[0]
+                        D2 = (X_sl[missing_index_sl, np.newaxis] - statistics) ** 2
```

Comment:
the travis error is because ``X_sl`` is already 2d, so you have to pass two indices in the older numpy. You want it to be 3d, right? Then it should be ``X_sl[missing_index_sl, :, np.newaxis]``.